File: TigerConnect Demo/TigerConnect Demo/Cells/OrganizationCell.m
//
//  OrganizationCell.m
//  TigerConnect Demo
//
//  Created by Oren Zitoun on 11/20/15.
//  Copyright © 2015 TigerText. All rights reserved.
//

#import "OrganizationCell.h"

@interface OrganizationCell ()

@property (weak, nonatomic) IBOutlet UILabel *organizationNameLabel;
@property (weak, nonatomic) IBOutlet UIView *selectedIndicatorView;
@property (weak, nonatomic) IBOutlet UIView *unreadBadgeView;
@property (weak, nonatomic) IBOutlet UILabel *unreadBadgeCountLabel;
@property (nonatomic) TTBadgeData *badgeData;

@end

@implementation OrganizationCell

+ (CGFloat)cellHeight
{
    return 60.0;
}

- (void)awakeFromNib
{
    self.selectedIndicatorView.backgroundColor = leadingColor();
}

- (void)setItem:(TTOrganization *)item
{
    _item = item;
    self.organizationNameLabel.text = _item.name;
    UIFont *font = nil;
    UIColor *textColor = nil;
    if ([_item.token isEqualToString:[[TTKit sharedInstance] currentOrganizationToken]]) {
        font = mediumFont(self.organizationNameLabel.font.pointSize);
        textColor = [UIColor blackColor];
        self.backgroundColor = [UIColor colorFromRGB:0xFDEBA9];
        self.selectedIndicatorView.hidden = NO;
    }
    else {
        font = regularFont(self.organizationNameLabel.font.pointSize);
        textColor = defaultGrayColor();
        self.backgroundColor = [UIColor whiteColor];
        self.selectedIndicatorView.hidden = YES;
    }
    [self setUnreadBadge];
    self.organizationNameLabel.font = font;
    self.organizationNameLabel.textColor = textColor;
}

// Setting the unread messages count badge
- (void)setUnreadBadge
{
    self.badgeData = [[TTKit sharedInstance] badgeDataForOrganizationToken:_item.token];
    [self updateUnreadCount:self.badgeData.unreadCount];
    
    __weak __typeof(self) weakSelf = self;
    self.badgeData.updateUnreadCountBlock = ^void(NSUInteger unreadCount) {
        [weakSelf updateUnreadCount:unreadCount];
    };
}

- (void)updateUnreadCount:(NSUInteger)unreadMessagesCountForConversation
{
    if (unreadMessagesCountForConversation > 0) {
        self.unreadBadgeView.hidden = NO;
        self.unreadBadgeCountLabel.text = [NSString stringWithFormat:@"%ld",(long)unreadMessagesCountForConversation];
    } else {
        self.unreadBadgeView.hidden = YES;
        self.unreadBadgeCountLabel.text = @"";
    }
    
    self.unreadBadgeView.layer.cornerRadius = MAX(CGRectGetHeight(self.unreadBadgeView.frame), 19)/2;
    [self.contentView setNeedsLayout];
    [self.contentView layoutIfNeeded];
}

- (void)layoutSubviews
{
    [super layoutSubviews];
    
    [self.unreadBadgeView setNeedsLayout];
    [self.unreadBadgeView layoutIfNeeded];
    self.unreadBadgeView.layer.cornerRadius = MAX(CGRectGetHeight(self.unreadBadgeView.frame), 19)/2;
}

- (void)prepareForReuse
{
    [self.badgeData prepareForReuse];
    self.unreadBadgeView.hidden = YES;
    self.unreadBadgeCountLabel.text = @"";
    [super prepareForReuse];
}

@end

File: TC Swift Demo/Cells/OrganizationCell.swift
//
//  OrganizationCell.swift
//  TC Swift Demo
//
//  Created by Oren Zitoun on 2/12/16.
//  Copyright © 2016 TigerConnect. All rights reserved.
//

import Foundation
// FIXME: comparison operators with optionals were removed from the Swift Standard Libary.
// Consider refactoring the code to use the non-optional operators.
fileprivate func < <T : Comparable>(lhs: T?, rhs: T?) -> Bool {
  switch (lhs, rhs) {
  case let (l?, r?):
    return l < r
  case (nil, _?):
    return true
  default:
    return false
  }
}

// FIXME: comparison operators with optionals were removed from the Swift Standard Libary.
// Consider refactoring the code to use the non-optional operators.
fileprivate func > <T : Comparable>(lhs: T?, rhs: T?) -> Bool {
  switch (lhs, rhs) {
  case let (l?, r?):
    return l > r
  default:
    return rhs < lhs
  }
}


class OrganizationCell: UITableViewCell {

    @IBOutlet weak var organizationNameLabel: UILabel!
    @IBOutlet weak var unreadBadgeView: UIView!
    @IBOutlet weak var unreadBadgeCountLabel: UILabel!
    @IBOutlet weak var selectedIndicatorView: UIView!
    var badgeData: TTBadgeData?
    
    class func cellHeight() -> CGFloat {
        return 60.0
    }
    
    override func awakeFromNib() {
        super.awakeFromNib()
        self.selectedIndicatorView.backgroundColor = Constants.Colors.leadingColor
    }
    
    var item: TTOrganization? {
        didSet {
            self.organizationNameLabel.text = item?.name
            let font: UIFont!
            let textColor: UIColor!
            if item?.token == TTKit.sharedInstance().currentOrganizationToken() {
                font = Constants.Fonts.mediumFont(self.organizationNameLabel.font.pointSize)
                textColor = UIColor.black
                self.backgroundColor = UIColor.init(hex: 0xFDEBA9)
                self.selectedIndicatorView.isHidden = false
            } else {
                font = Constants.Fonts.regularFont(self.organizationNameLabel.font.pointSize)
                textColor = Constants.Colors.defaultGrayColor
                self.backgroundColor = UIColor.white
                self.selectedIndicatorView.isHidden = true
            }
            self.setUnreadBadge()
            self.organizationNameLabel.font = font
            self.organizationNameLabel.textColor = textColor
        }
    }
    
    fileprivate func setUnreadBadge() {
        self.badgeData = TTKit.sharedInstance().badgeData(forOrganizationToken: item?.token)
        self.updateUnreadCount(self.badgeData?.unreadCount)
        
    }
    
    fileprivate func updateUnreadCount(_ unreadMessagesCountForConversation: UInt?) {
        if unreadMessagesCountForConversation > 0 {
            self.unreadBadgeView.isHidden = false
            self.unreadBadgeCountLabel.text = String.init(format: "%ld", unreadMessagesCountForConversation!)
        } else {
            self.unreadBadgeView.isHidden = true
            self.unreadBadgeCountLabel.text = ""
        }
        self.unreadBadgeView.cornerRadius = max(self.unreadBadgeView.frame.height, 19)/2
        self.contentView.setNeedsLayout()
        self.contentView.layoutIfNeeded()
    }
    
    override func layoutSubviews() {
        super.layoutSubviews()
        self.unreadBadgeView.setNeedsLayout()
        self.unreadBadgeView.layoutIfNeeded()
        self.unreadBadgeView.cornerRadius = max(self.unreadBadgeView.frame.height, 19)/2
    }
    
    override func prepareForReuse() {
        super.prepareForReuse()
        self.badgeData?.prepareForReuse()
        self.unreadBadgeView.isHidden = true
        self.unreadBadgeCountLabel.text = ""
    }
}
